### ISSUE TITLE
Timeout issue

### ISSUE THREAD

**Issue by peetasan** [awslabs/aws-lambda-redshift-loader#91]
I set up my Lambda according to your docs but I got the following in the logs, with 15 seconds timeout (for testing with sample data):

```
2016-07-06T13:08:39.478Z    c12cab70-437a-11e6-a60e-374242d5f8b1    Found Redshift Load Configuration for ***/sample 
2016-07-06T13:08:39.593Z    c12cab70-437a-11e6-a60e-374242d5f8b1    Adding Pending Batch Entry for ***/sample/sample-redshift-file-for-lambda-loader2.csv 
2016-07-06T13:08:39.742Z    c12cab70-437a-11e6-a60e-374242d5f8b1    Batch Size 1 reached 
2016-07-06T13:08:39.742Z    c12cab70-437a-11e6-a60e-374242d5f8b1    Batch Size 1 not reached but reached Age 120 seconds 
2016-07-06T13:08:39.880Z    c12cab70-437a-11e6-a60e-374242d5f8b1    Creating Manifest for Batch 0e1e5994-6956-41a0-b73c-f0f6c08f85ac 
2016-07-06T13:08:39.881Z    c12cab70-437a-11e6-a60e-374242d5f8b1    Writing manifest to ***/manifests/manifest-2016-07-06-13:08:39-9605 
2016-07-06T13:08:39.999Z    c12cab70-437a-11e6-a60e-374242d5f8b1    Created Manifest ***/manifests/manifest-2016-07-06-13:08:39-9605 Successfully 
2016-07-06T13:08:40.003Z    c12cab70-437a-11e6-a60e-374242d5f8b1    Loading 1 Clusters 
2016-07-06T13:08:40.004Z    c12cab70-437a-11e6-a60e-374242d5f8b1    Remaining duration of 4293 insufficient to load cluster 
2016-07-06T13:08:40.010Z    c12cab70-437a-11e6-a60e-374242d5f8b1    Loading Cluster ***.redshift.amazonaws.com with Lambda credentials 
2016-07-06T13:08:40.082Z    c12cab70-437a-11e6-a60e-374242d5f8b1    Linked Manifest manifest-2016-07-06-13:08:39-9605 to Batch 0e1e5994-6956-41a0-b73c-f0f6c08f85ac 
2016-07-06T13:08:40.098Z    c12cab70-437a-11e6-a60e-374242d5f8b1    Created new Failed Manifest ***/failed_loads/manifest-2016-07-06-13:08:39-9605 
2016-07-06T13:08:40.250Z    c12cab70-437a-11e6-a60e-374242d5f8b1    begin; COPY sample from 's3://***/manifests/manifest-2016-07-06-13:08:39-9605' with credentials as 'aws_access_key_id=***;aws_secret_access_key=***;token=***' manifest format csv delimiter ',' ; commit; 
2016-07-06T13:08:40.250Z    c12cab70-437a-11e6-a60e-374242d5f8b1    Connecting to Database ***.redshift.amazonaws.com:5439 
2016-07-06T13:08:40.254Z    c12cab70-437a-11e6-a60e-374242d5f8b1    "Remaining duration of 4293 insufficient to load cluster" 
2016-07-06T13:08:40.254Z    c12cab70-437a-11e6-a60e-374242d5f8b1
{
    "errorMessage": "error"
}
END RequestId: c12cab70-437a-11e6-a60e-374242d5f8b1 
REPORT RequestId: c12cab70-437a-11e6-a60e-374242d5f8b1  Duration: 957.68 ms Billed Duration: 1000 ms Memory Size: 1024 MB   Max Memory Used: 25 MB
```

Here, I increased the timeout to 1 minute, also deleted the items from the LambdaRedshiftBatches and 
LambdaRedshiftProcessedFiles tables, and the manifest and failed load files from S3.
Since then I get `Task timed out after 60 seconds`. (I tried with 2 and 5 minutes too, same result). Also recreated the Dynamo config, the table in Redshift, added some logs to the code itself, but nothing seems to help. 

I am using the 2.4.0 dist, my Lambda's security group is allowed to reach Redshift on port 5439. I have the required policies for the IAM Role. What else could I miss? Thanks.

**Comment by IanMeyers** [comment]
it looks like your lambda invocation is never connecting. Is your function configured to run in a VPC, or just as standard. If standard, have you opened up 5439 connections from 0.0.0.0/0 so that lambda can connect?

**Comment by peetasan** [comment]
It is running in a VPC, the same one which the Redshift cluster is in.

**Comment by IanMeyers** [comment]
When you say that Lambda's security group is allowed to reach Redshift, can I just confirm that you mean that you've added a Security Group Grant in the Cluster Security Group for your Cluster to the Lambda function's Security Group?

**Comment by peetasan** [comment]
I can't use Cluster security groups, only VPC security groups for my account.

I have a security group (redshift), which allows access from another security group (lambda) on port 5439. I have added the latter on Lambda - Configuration - Advanced settings as the security group my Lambda function should use. 

I also set up a NAT in the VPC the Redshift is in, though this did not help (also it would make other clients unable to access Redshift).

I am also unsure why I could at some point connect to at least S3 but not anymore.

Thanks

**Comment by IanMeyers** [comment]
OK - am going to build this configuration in my own account and try to see where the issue might be.

Ian

**Comment by IanMeyers** [comment]
I have updated the top level documentation for this project with the details of how I was able to use a VPC configuration to connect to my Redshift cluster and load data. Can you review and then ensure you are using the same configuration? If so, then let me know and we'll address.

**Comment by peetasan** [comment]
Based on the new documentation it works! Thanks a lot for the help.

Here's what I did:
I created 4 subnets: 2 public, 2 private, one in each AZ. I have created 2 NATs in the 2 private subnets respectively. I have created 2 route tables 1 for each private subnet adding the two NATs to them respectively. I have created a 3rd route table for the 2 public subnets with an Internet Gateway. I have also set Lambda to use the private subnets. I didn't have to change the security group settings I already had.
